feat(headers): Implement HSTS preload

gabibbo97 opened the pull request:
Prior to this, Hyper did not implement the HSTS preload header,
this commit changes the header in order to implement a constructor that verifies the validity of max-age at compile time, a coherent string to header parser and some tests to ensure functionality.

This closes issue  #1334 
**BREAKING CHANGE:** the StrictTransportSecurity struct now has a new bool member: preload

- [x] The commit messages match the guidelines in https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md#git-commit-guidelines

seanmonstar commented:
Cool! Since you're in there, doing all the breaks, how would you feel making the fields all private, and have setters/getters instead? That'd allow any new fields to be added without breaking changes in the future...

gabibbo97 commented:
I have made a significant breaking change to the header:

- all members are now private
- you can create the header using a builder-like pattern

I have updated also the documentation
This is my first pull-request of all time in Rust 😄

gabibbo97 commented:
I did something wrong with my understanding of the "bench_headers" macro

gabibbo97 commented:
Please wait, currently I am refactoring the tests

gabibbo97 commented:
Everything should be ok now

seanmonstar commented on src/header/common/strict_transport_security.rs:
```diff
@@ -39,43 +40,108 @@ use header::{Header, Raw, parsing};
+    pub fn max_age(&self) -> u64 { self.max_age }
+
+    /// Sets the max age
+    pub fn set_max_age(&self, max_age: u64) -> StrictTransportSecurity {
```
It's more usual for setters to just mutate the type itself, than to return a new type. We should probably stick with the usual :D

seanmonstar commented on src/header/common/strict_transport_security.rs:
```diff
@@ -39,43 +40,108 @@ use header::{Header, Raw, parsing};
 }
 
+/// This value indicates the minimum value of the max_age
+/// that can be used togheter with preload
```
Typo: "together".

seanmonstar commented on src/header/common/strict_transport_security.rs:
```diff
@@ -39,43 +40,108 @@ use header::{Header, Raw, parsing};
+    /// 
+    /// * It also applies to subdomains
+    /// * It has a max_age greater than 31536000 seconds
+    pub fn set_preload(&self, preload: bool) -> Result<StrictTransportSecurity,()> {
```
Hm, I see why this returns a `Result`. Google requires the max age minimum in order for a site to eligible to for its own preload list.

Though, I'm not sure if we should be enforcing that specific requirement here.

seanmonstar commented on src/header/common/strict_transport_security.rs:
```diff
@@ -85,35 +151,56 @@ impl FromStr for StrictTransportSecurity {
+                    preload: false
                 }),
+                (Some(age), sub, pre) if pre.is_some() => {
+                    if sub.is_none() || age < PRELOAD_MIN_AGE {
```
I wouldn't enforce the limit when parsing the header either. It's a perfectly valid header, it just doesn't meet the requirements for Google's list at the time of writing (but could in the future).

gabibbo97 commented:
Hi Sean, thanks for your review, from the last time I have edited all of the setters from builder-pattern like to proper setter methods.
I also agree with you on the preload issue, the RFC tells us that a StrictTransportSecurity header that does not comply with Google's guidelines is still valid so I have removed the check and added a method that checks if your current struct matches the guidelines, in this way I believe that we can manage to parse valid but not Google compliant headers and also check if our new header meets the requisites.
The docs have been updated according to this change

seanmonstar commented:
As noted in #189, headers are actually going to moved out of hyper proper. So as to not keep this dangling, I'm going to close this. Thanks for the submission, and of course, because the header system works for anything that implements `Header`, anyone can use these changes themselves by creating a new type.
